pydm/widgets/archiver_time_plot.py

```diff
@@ -286,6 +292,313 @@ def receiveNewValue(self, new_value):
+
+    def __init__(
+        self,
+        formula: str = None,
+        pvs: dict = None,
+        use_archive_data: Optional[bool] = True,
+        liveData: Optional[bool] = True,
+        color: Optional[str] = "green",
+        **kws
+    ):
+        super(FormulaCurveItem, self).__init__(**kws)
+        self.color = color
+        self.use_archive_data = use_archive_data
+        self.archive_points_accumulated = 0
+        # Start with empty buffers because we don't
+        # calculate anything until we try to draw the curve
+        self._archiveBufferSize = DEFAULT_ARCHIVE_BUFFER_SIZE
+        self._bufferSize = 0
+        self.archive_data_buffer = np.zeros((2, 0), order="f", dtype=float)
+
+        self.data_buffer = np.zeros((2, 0), order="f", dtype=float)
+
+        # When optimized or mean value data is requested, we can display error bars representing
+        # the full range of values retrieved
+        self.error_bar_item = ErrorBarItem()
+        self.error_bar_needs_set = True
+        self._formula = formula
+        # Have a formula for internal calculations, that the user does not see
+        self._trueFormula = self.createTrueFormula()
+        self.minx = float("-inf")
+        self.maxx = float("inf")
+        self.pvs = pvs
+        self._liveData = liveData
+        self.plot_style = "Line"
+
+    def to_dict(self) -> OrderedDict:
+        """Returns an OrderedDict representation with values for all properties needed to recreate this curve."""
+        dic_ = OrderedDict([("useArchiveData", self.use_archive_data), ("liveData", self.liveData)])
+        dic_.update(super(ArchivePlotCurveItem, self).to_dict())
+        return dic_
+
+    @property
+    def liveData(self):
+        for pv in self.pvs.keys():
+            if not self.pvs[pv].liveData:
+                return False
+        return True
+
+    @liveData.setter
+    def liveData(self, get_live: bool):
+        if not get_live:
+            self._liveData = False
+            return
+        self._liveData = True
+
+    @property
+    def formula(self):
+        return self._formula
+
+    @formula.setter
+    def formula(self, formula: str):
+        self._formula = formula
+        self._trueFormula = self.createTrueFormula()
+
+    def checkFormula(self) -> bool:
+        """Confirm that our formula is still valid. Namely, all of the curves we depend on are still in use"""
+        for pv in self.pvs.keys():
+            if not self.pvs[pv].exists:
+                print(pv + " is no longer a valid row name")
+                # If one of the rows we rely on is gone, not only are we no longer a valid formula,
+                # but all rows that rely on us are also invalid.
+                self.exists = False
+                return False
+        return True
+
+    def createTrueFormula(self) -> str:
+        """Convert our human-readable formula to something easier to use for the computer, in the background only"""
+        formula = self.formula[4:]
```

Comment:
I've added these changes; however, I have it set up on the Archive Viewer repo side that a formula is created iff the channel starts with f://, so such a check *shouldn't* be necessary unless somebody else is using the FormulaCurveItem object. Because the point of this is that potentially somebody else does, I think it makes sense regardless